README
Kosmonaut - The WebRocket client toolkit
========================================

The 'kosmonaut' library is a toolkit to build efficient clients for
the WebRocket link:http://rfc.webrocket.io/draft/BWP.html[*Backend
Workers Protocol*].

IMPORTANT: This is not a full stack client, it's just a low level
	framework which helps to build clients in other programming
	languages without dealing with ZeroMQ magic and quirks.

Installation
------------
The file 'INSTALL' in this directory says how to build and install 
Kosmonaut on various systems, once you have unpacked or checked out the
entire Kosmonaut's file tree.

Release history
---------------
See the file 'NEWS' for information on new features and other user-visible
changes in recent versions of Kosmonaut.

Contributing
------------
We appreciate any kind of help in Kosmonaut development and testing.
If you don't know how can you help us, please check the project's
issues or 'TODO' file in project root directory. Before you start
working on the code, and before you send us a pull request, make sure
you understand and follow the contributors guide, which you can find
in the 'CONTRIBUTE' file.

Copyright
---------
Copyright (C) 2011 by Krzysztof Kowalik <chris@nu7hat.ch>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.


NEWS
Release history
===============

v0.0.1
------
* Client library implementing the WebRocket's Backend Worker Protocol
  low level interface.

src/astronaut.c
#include <czmq.h>
#include <stdlib.h>
#include <uuid/uuid.h>
#include <string.h>
#include <unistd.h>

#include "astronaut.h"

#define SA_TOKEN_LEN 128
#define UUID_LEN 36
#define CMD_LEN 4
#define REQUEST_TIMEOUT  1000
#define RESPONSE_TIMEOUT 5000
#define LISTENER_TIMEOUT 1000

astronaut_t* astronaut_new(const char* addr, const char *vhost, const char *secret)
{
	astronaut_t* self;
	uuid_t uuid;
	int addr_len = strlen(addr);
	char* sid = (char*)malloc(UUID_LEN * sizeof(char));
	char identity[strlen(addr) + strlen(vhost) + strlen(secret) + 3 + 3];
  
	self = (astronaut_t*)malloc(sizeof(astronaut_t));
	if (!self)
		return NULL;

	pthread_mutex_init(&self->lmtx, NULL);
	pthread_mutex_init(&self->tmtx, NULL);

	self->running = 0; // not running, ready to run
  
	self->addr = (char*)malloc(addr_len * sizeof(char));
	memcpy(self->addr, addr, addr_len * sizeof(char));
	
	uuid_generate(uuid);
	uuid_unparse_lower(uuid, sid);
  
	self->ctx = zctx_new();
	self->req = zsocket_new(self->ctx, ZMQ_REQ);
	sprintf(identity, "req:%s:%s:%s", vhost, secret, sid);
	zsockopt_set_identity(self->req, identity);

	self->dlr = zsocket_new(self->ctx, ZMQ_DEALER);
	sprintf(identity, "dlr:%s:%s:%s", vhost, secret, sid);
	zsockopt_set_identity(self->req, identity);
  
	free(sid);

	// TODO: register signal handlers
  
	return self;
}

void astronaut_destroy(astronaut_t** self_p)
{
	if (!*self_p)
		return;
  
	astronaut_t* self = *self_p;
  
	astronaut_stop_listening(self);
	zctx_destroy(&self->ctx);
	pthread_mutex_destroy(&self->lmtx);
	pthread_mutex_destroy(&self->tmtx);
	free(self);
	*self_p = NULL;
}

int astronaut_connect(astronaut_t* self)
{
	return zsocket_connect(self->req, self->addr);
}

zmsg_t* astronaut_request(astronaut_t* self, char* request)
{
	if (!request || !self)
		return NULL;

	zmq_pollitem_t items[] = {{self->req, 0, ZMQ_POLLOUT|ZMQ_POLLIN, 0}};
	zmsg_t *res = NULL, *req = NULL;

	pthread_mutex_lock(&self->tmtx);

	int rc = zmq_poll(items, 1, REQUEST_TIMEOUT * ZMQ_POLL_MSEC);
	if (rc == 0 && items[0].revents & ZMQ_POLLOUT) {
		zmsg_t *req = zmsg_new();
		zmsg_addstr(req, request);
		zmsg_send(&req, self->req);
		zmsg_destroy(&req);
	
		rc = zmq_poll(items, 1, RESPONSE_TIMEOUT * ZMQ_POLL_MSEC);
		if (rc == 0 && items[0].revents & ZMQ_POLLIN) {
			res = zmsg_recv(self->req);
		}
	}
  
	pthread_mutex_lock(&self->tmtx);
	return res;
}

char* astronaut_get_single_access_token(astronaut_t* self, const char* perm)
{
	if (!perm || !self)
		return NULL;
  
	zmsg_t* res = NULL;
	char* token = NULL;
	char* cmd = (char*)malloc((strlen(perm) + CMD_LEN) * sizeof(char));

	sprintf(cmd, "SAT %s", perm);
	res = astronaut_request(self, cmd);

	if (!res)
		return NULL;
	  
	token = zmsg_popstr(res);
	free(cmd);

	return token;
}

int astronaut_trigger(astronaut_t* self, char* data)
{
	if (!data || !self)
		return -1;

	int rc;
	char* rcs = NULL;
	zmsg_t* res = NULL;
	char* payload = (char*)malloc((strlen(data) + CMD_LEN) * sizeof(char));

	sprintf(payload, "TRG %s", data);
	res = astronaut_request(self, data);

	if (!res)
		return -1;

	rcs = zmsg_popstr(res);
	sscanf(rcs, "%d", &rc);
	return rc;
}

int astronaut_listen(astronaut_t* self, astronaut_listener_t callback)
{
	if (!self)
		return -1;
	if (self->running == 1)
		return E_RUNNING; // is running
  
	int rc = zsocket_connect(self->dlr, self->addr);
	if (rc != 0)
		return rc;

	char* payload = NULL;
	zmsg_t* res = NULL;
	zmq_pollitem_t items[] = {{self->req, 0, ZMQ_POLLIN, 0}};

	self->running = 1; // running
  
	while (1) {
		if (self->running < 1) { // stop
			break;
		}
	
		int rc = zmq_poll(items, 1, LISTENER_TIMEOUT * ZMQ_POLL_MSEC);
		if (rc == -1)
			continue;

		if (items[0].revents & ZMQ_POLLIN) {
			res = zmsg_recv(self->req);
			payload = zmsg_popstr(res);
			callback(self, payload);
		}
	}

	self->running = -1; // stopped
	zmsg_destroy(&res);
	return 0;
}

void astronaut_stop_listening(astronaut_t* self)
{
	if (!self || self->running < 1)
		return;
  
	pthread_mutex_lock(&self->lmtx);
	self->running = 0;
	while (self->running != -1)
		usleep(1000);
	pthread_mutex_unlock(&self->lmtx);
}


include/astronaut.h
#ifndef _ASTRONAUT_H_
#define _ASTRONAUT_H_

#include <pthread.h>

#define E_RUNNING 101

/**
 * Astronaut representation structure.
 */
typedef struct astronaut_t {
	zctx_t* ctx;
	void* req;
	void* dlr;
	char* addr;
	short int running;
	pthread_mutex_t lmtx; // listener mutex
	pthread_mutex_t tmtx; // trigger mutex
} astronaut_t;

/**
 * Listener's callback function type. 
 */
typedef void(*astronaut_listener_t)(astronaut_t*, char*);

/**
 * Astronaut constructor.
 * Creates new instance of the `astronaut_t` and initializes it's
 * internal sockets. The `astronaut_t` contains 2 kind of sockets:
 *
 * - ZMQ_REQ: handles authentication and tasks triggering
 * - ZMQ_DEALER: handles incoming tasks
 *
 * Memory allocated by `astronaut_new` have to be freed using the
 * `astronaut_destroy` function.
 *
 * @see astronaut_destroy
 * @see astronaut_connect
 *
 * @param addr   the server URI address
 * @param vhost  vhost to connect to
 * @param secret secret key assigned to given vhost
 * @return Configured astronaut instance
 */
astronaut_t* astronaut_new(const char* addr, const char* vhost, const char* secret);

/**
 * Astronaut destructor.
 * Stops event loops, disconnects the sockets and cleans up the context.
 * Frees all allocated memory as well.
 *
 * @param self_p memory address of an astronaut instance.
 */
void astronaut_destroy(astronaut_t** self_p);

/**
 * The REQ socket connector.
 * Establishes connection for the REQ socket only. To establish
 * connection for the DEALER the `astronaut_listen` function needs
 * to be used.
 *
 * @see astronaut_listen
 *
 * @param self an astronaut instance
 * @param uri  the URI to connect to
 * @return 0 if OK, otherwise appropriate error code
 */
int astronaut_connect(astronaut_t* self);

/**
 * Request for signle access token.
 * Gets a single access token from the WebRocket. Single access token
 * is used to authenticate the frontend client in case of using the
 * private channels.
 *
 * The single access token is generated for specified permissions -
 * which means each token can be used to access only those channels
 * matching the permission regexp.
 *
 * @param self an astronaut instance
 * @param perm permission regexp
 * @return a single access token string
 */
char* astronaut_get_single_access_token(astronaut_t* self, const char* perm);

/**
 * Trigger an WebRocket operation.
 * Sends trigger message to the WebRocket backend and waits for
 * confirmation.
 *
 * @param self an astronaut instance
 * @param data serialized data with trigger command (JSON-encoded string)
 * @return 0 if OK, otherwise appropriate error code
 */
int astronaut_trigger(astronaut_t* self, char* data);

/**
 * Starts the listener's event loop.
 * Establishes the DEALER connection and starts listener event loop for
 * it. Listener can't be activated more than once.
 *
 * @param self     an astronaut instance
 * @param callback callback function called when a message is received.
 * @return 0 if OK, otherwise appropriate error code
 */
int astronaut_listen(astronaut_t* self, astronaut_listener_t callback);

/**
 * Stops the listener.
 * If a listener's event loop is running, then stops it and cleans up
 * its status.
 *
 * @param self an astronaut instance
 */
void astronaut_stop_listening(astronaut_t* self);

#endif /* _ASTRONAUT_H_ */


include/kosmonaut.h
#ifndef _KOSMONAUT_H_
#define _KOSMONAUT_H_

#include <pthread.h>

#define E_RUNNING 101

/**
 * Kosmonaut representation structure.
 */
typedef struct kosmonaut_t {
	zctx_t* ctx;
	void* req;
	void* dlr;
	char* addr;
	short int running;
	pthread_mutex_t lmtx; // listener mutex
	pthread_mutex_t tmtx; // trigger mutex
} kosmonaut_t;

/**
 * Listener's callback function type. 
 */
typedef void(*kosmonaut_listener_t)(kosmonaut_t*, char*);

/**
 * Kosmonaut constructor.
 * Creates new instance of the `kosmonaut_t` and initializes it's
 * internal sockets. The `kosmonaut_t` contains 2 kind of sockets:
 *
 * - ZMQ_REQ: handles authentication and tasks triggering
 * - ZMQ_DEALER: handles incoming tasks
 *
 * Memory allocated by `kosmonaut_new` have to be freed using the
 * `kosmonaut_destroy` function.
 *
 * @see kosmonaut_destroy
 * @see kosmonaut_connect
 *
 * @param addr   the server URI address
 * @param vhost  vhost to connect to
 * @param secret secret key assigned to given vhost
 * @return Configured kosmonaut instance
 */
kosmonaut_t* kosmonaut_new(const char* addr, const char* vhost, const char* secret);

/**
 * Kosmonaut destructor.
 * Stops event loops, disconnects the sockets and cleans up the context.
 * Frees all allocated memory as well.
 *
 * @param self_p memory address of an kosmonaut instance.
 */
void kosmonaut_destroy(kosmonaut_t** self_p);

/**
 * The REQ socket connector.
 * Establishes connection for the REQ socket only. To establish
 * connection for the DEALER the `kosmonaut_listen` function needs
 * to be used.
 *
 * @see kosmonaut_listen
 *
 * @param self an kosmonaut instance
 * @param uri  the URI to connect to
 * @return 0 if OK, otherwise appropriate error code
 */
int kosmonaut_connect(kosmonaut_t* self);

/**
 * Request for signle access token.
 * Gets a single access token from the WebRocket. Single access token
 * is used to authenticate the frontend client in case of using the
 * private channels.
 *
 * The single access token is generated for specified permissions -
 * which means each token can be used to access only those channels
 * matching the permission regexp.
 *
 * The memory of returned access token have to be freed by the user.
 * 
 * @param self an kosmonaut instance
 * @param perm permission regexp
 * @return a single access token string
 */
char* kosmonaut_get_single_access_token(kosmonaut_t* self, const char* perm);

/**
 * Trigger an WebRocket operation.
 * Sends trigger message to the WebRocket backend and waits for
 * confirmation.
 *
 * @param self an kosmonaut instance
 * @param data serialized data with trigger command (JSON-encoded string)
 * @return 0 if OK, otherwise appropriate error code
 */
int kosmonaut_trigger(kosmonaut_t* self, char* data);

/**
 * Starts the listener's event loop.
 * Establishes the DEALER connection and starts listener event loop for
 * it. Listener can't be activated more than once.
 *
 * The memory of the message data passed to the callback have to be
 * freed by the user.
 *
 * @param self     an kosmonaut instance
 * @param callback callback function called when a message is received.
 * @return 0 if OK, otherwise appropriate error code
 */
int kosmonaut_listen(kosmonaut_t* self, kosmonaut_listener_t callback);

/**
 * Stops the listener.
 * If a listener's event loop is running, then stops it and cleans up
 * its status.
 *
 * @param self an kosmonaut instance
 */
void kosmonaut_stop_listening(kosmonaut_t* self);

#endif /* _KOSMONAUT_H_ */


src/kosmonaut.c
#include <czmq.h>
#include <stdlib.h>
#include <uuid/uuid.h>
#include <string.h>
#include <unistd.h>

#include "kosmonaut.h"

#define SA_TOKEN_LEN 128
#define UUID_LEN 36
#define CMD_LEN 4
#define REQUEST_TIMEOUT  1000
#define RESPONSE_TIMEOUT 5000
#define LISTENER_TIMEOUT 1000

kosmonaut_t* kosmonaut_new(const char* addr, const char *vhost, const char *secret)
{
	kosmonaut_t* self;
	uuid_t uuid;
	int addr_len = strlen(addr);
	char* sid = (char*)malloc(UUID_LEN * sizeof(char));
	char identity[strlen(addr) + strlen(vhost) + strlen(secret) + 3 + 3];
  
	self = (kosmonaut_t*)malloc(sizeof(kosmonaut_t));
	if (!self)
		return NULL;

	pthread_mutex_init(&self->lmtx, NULL);
	pthread_mutex_init(&self->tmtx, NULL);

	self->running = 0; // not running, ready to run
  
	self->addr = (char*)malloc(addr_len * sizeof(char));
	memcpy(self->addr, addr, addr_len * sizeof(char));
	
	uuid_generate(uuid);
	uuid_unparse_lower(uuid, sid);
  
	self->ctx = zctx_new();
	self->req = zsocket_new(self->ctx, ZMQ_REQ);
	sprintf(identity, "req:%s:%s:%s", vhost, secret, sid);
	zsockopt_set_identity(self->req, identity);

	self->dlr = zsocket_new(self->ctx, ZMQ_DEALER);
	sprintf(identity, "dlr:%s:%s:%s", vhost, secret, sid);
	zsockopt_set_identity(self->req, identity);
  
	free(sid);

	// TODO: register signal handlers
  
	return self;
}

void kosmonaut_destroy(kosmonaut_t** self_p)
{
	if (!*self_p)
		return;
  
	kosmonaut_t* self = *self_p;
  
	kosmonaut_stop_listening(self);
	zctx_destroy(&self->ctx);
	pthread_mutex_destroy(&self->lmtx);
	pthread_mutex_destroy(&self->tmtx);
	free(self);
	*self_p = NULL;
}

int kosmonaut_connect(kosmonaut_t* self)
{
	return zsocket_connect(self->req, self->addr);
}

zmsg_t* kosmonaut_request(kosmonaut_t* self, char* request)
{
	if (!request || !self)
		return NULL;

	zmq_pollitem_t items[] = {{self->req, 0, ZMQ_POLLOUT|ZMQ_POLLIN, 0}};
	zmsg_t* res = NULL;
	zmsg_t* req = NULL;

	pthread_mutex_lock(&self->tmtx);

	int rc = zmq_poll(items, 1, REQUEST_TIMEOUT * ZMQ_POLL_MSEC);
	if (rc == 0 && items[0].revents & ZMQ_POLLOUT) {
		req = zmsg_new();
		zmsg_addstr(req, request);
		zmsg_send(&req, self->req);
		zmsg_destroy(&req);
	
		rc = zmq_poll(items, 1, RESPONSE_TIMEOUT * ZMQ_POLL_MSEC);
		if (rc == 0 && items[0].revents & ZMQ_POLLIN) {
			res = zmsg_recv(self->req);
		}
	}
  
	pthread_mutex_lock(&self->tmtx);
	return res;
}

char* kosmonaut_get_single_access_token(kosmonaut_t* self, const char* perm)
{
	if (!perm || !self)
		return NULL;
  
	zmsg_t* res = NULL;
	char* token = NULL;
	char* cmd = (char*)malloc((strlen(perm) + CMD_LEN) * sizeof(char));

	sprintf(cmd, "SAT %s", perm);
	res = kosmonaut_request(self, cmd);

	if (!res)
		return NULL;
	  
	token = zmsg_popstr(res);
	free(cmd);

	return token;
}

int kosmonaut_trigger(kosmonaut_t* self, char* data)
{
	if (!data || !self)
		return -1;

	int rc;
	char* rcs = NULL;
	zmsg_t* res = NULL;
	char* payload = (char*)malloc((strlen(data) + CMD_LEN) * sizeof(char));

	sprintf(payload, "TRG %s", data);
	res = kosmonaut_request(self, data);

	if (!res)
		return -1;

	rcs = zmsg_popstr(res);
	sscanf(rcs, "%d", &rc);
	return rc;
}

int kosmonaut_listen(kosmonaut_t* self, kosmonaut_listener_t callback)
{
	if (!self)
		return -1;
	if (self->running == 1)
		return E_RUNNING; // is running
  
	int rc = zsocket_connect(self->dlr, self->addr);
	if (rc != 0)
		return rc;

	zmsg_t* res = NULL;
	zmsg_t* hbt = NULL;
	char* payload = NULL;
	zmq_pollitem_t items[] = {{self->req, 0, ZMQ_POLLIN, 0}};

	self->running = 1; // running
  
	while (1) {
		if (self->running < 1) { // stop
			break;
		}
	
		int rc = zmq_poll(items, 1, LISTENER_TIMEOUT * ZMQ_POLL_MSEC);
		if (rc == -1)
			continue;

		if (items[0].revents & ZMQ_POLLIN) {
			res = zmsg_recv(self->req);
			payload = zmsg_popstr(res);
			
			if (strncmp(payload, "HBT", 3) == 0) {
				// Handle heartbeat
				if (!hbt) {
					hbt = zmsg_new();
					zmsg_addstr(hbt, "HBT");
				}
				zmsg_send(&hbt, self->req);
			} else {
				// Process receieved message in the callback
				callback(self, payload);
			}
		}
	}

	self->running = -1; // stopped
	zmsg_destroy(&res);
	zmsg_destroy(&hbt);
	return 0;
}

void kosmonaut_stop_listening(kosmonaut_t* self)
{
	if (!self || self->running < 1)
		return;
  
	pthread_mutex_lock(&self->lmtx);
	self->running = 0;
	while (self->running != -1)
		usleep(1000);
	pthread_mutex_unlock(&self->lmtx);
}
